include/threepp/core/EventDispatcher.hpp

```diff
@@ -7,38 +7,115 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <functional>
+#include <algorithm>
+#include <atomic>
 
 
 namespace threepp {
 
-    struct Event {
+    namespace concepts {
+#if defined(__cpp_concepts) && (__cpp_concepts >= 201907L) 
+        template<typename TEvent>
+        concept Event = requires(TEvent e) {
+            { e.target } -> std::convertible_to<void*>;
+            { e.unsubscribe } -> std::convertible_to<bool>;
+        };
+#endif
+    }
+
+    /// An event listener is just a function that takes an argument of type TEvent
+    template<typename TEvent>
+    using TEventListener = std::function<void(TEvent&)>;
+
+    /// A single subscription for an event
+    using Subscription = std::shared_ptr<void>;
+
+    /// For holding a large number of subscriptions to events
+    using Subscriptions= std::vector<Subscription>;
+
+    /// Allows one to use the << to push subscriptions onto the vector
+    inline
+    void operator<<(std::vector<Subscription>& subs, Subscription const & sub) {
+        subs.push_back(sub);
+    }
+
+    /// Generic event dispatch class
+    template <typename TEvent>
+#if defined(__cpp_concepts) && (__cpp_concepts >= 201907L) 
+    requires concepts::Event<TEvent>
+#endif
+    // C++20 (and later) code
+    class TEventDispatcher {
+    public:
+        using EventListener = TEventListener<TEvent>;
+
+        /// Adds an event listener and returns a subscription
+        [[nodiscard]] Subscription subscribe(EventListener listener) {
+            size_t current_id = id_.load();
+            listeners_.insert({ current_id, listener });
+            Subscription disposer((void*) nullptr, [this, current_id](void*) { listeners_.erase(current_id); });
+            id_ = id_ + 1;
+            return disposer;
+        }
+
+        /// Adds an event listener and never automatically unsubscribes. You
+        /// can set event.unsubscribe = true and the subscription will be
+        /// cancelled. Not recommended to be used directly. Build other
+        /// tools on this.
+        void subscribeForever(EventListener listener) {
```

Comment:
Calling this function makes the application crash on shutdown as the natural order of destruction is changed. The same is true if a static Mesh is defined in the application, so not necessarly this PR's issue, but needs a fix.

Reply:
Not really an issue with this PR, and the underlying issue will be solved in a separate PR.